### PR TITLE
Update to use webpack-blocks 1.0.0 API

### PR DESCRIPTION
**Pull request by boxfoot**
This PR updates webpack-blocks-happypack to use the latest webpack-blocks API, and also provides support for webpack 3.

I think that it mostly works, but there are a few issues I'm not sure of, could use feedback:

- [x] After the change, the jest snapshots didn't match.  Some differences were obviously OK and I
  committed those - but there are a bunch where I can't tell if they're problems that need to be fixed or 
  unrelated changes in `webpack`, `webpack-blocks`, or other related libraries
- [x] I don't use yarn, so wasn't able to update `yarn.lock`, that still needs to be done if everything else
  checks out.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -75,9 +79,7 @@ const happypack = (
   return group(blocks.map(block =>
     Object.assign(happifyBlock(block, options), {
       pre: block.pre,
-      post: []
-        .concat(block.post || [])
-        .map(postHook => happifyBlock(postHook, options)),
+      post: block.post,
```
Comment (boxfoot) [inline src/index.js]: Wasn't clear what the old post hook processing was doing, or how to adapt it for new API, so I took it out.  @diegohaz happy to fix if you can explain the requirements.

File: src/index.js [hunk 2]
```diff
@@ -16,14 +16,18 @@ import {
...
   block: WebpackBlock,
   { loaders, ...happypackOptions }: BlockOptions
-): WebpackBlock => (...args): Block => {
-  const compiledBlock = block(...args)
+): WebpackBlock => (context, utils): Block => prevConfig => {
+  const compiledBlock = block(context, utils)(prevConfig)
   const originalRules = getRules(compiledBlock)
 
-  if (!originalRules) return compiledBlock
+  if (!originalRules) return prevConfig
```
Comment (boxfoot) [inline src/index.js]: Wasn't clear to me if this should return `prevConfig` or `compiledBlock` when there's nothing to change...

Comment (andywer) [reply src/index.js]: When there's nothing to change, return the unchanged `prevConfig`. It's fine 👍

File: src/index.js (hunk 1 above)
Comment (andywer) [reply src/index.js]: It's happifying the blocks' post hooks. I suppose that's got to stay like that:
```js
post: (block.post || []).map(postHook => happifyBlock(postHook, options))
```

But maybe @diegohaz knows more.

File: test/index.test.js [hunk 3]
```diff
@@ -35,17 +35,17 @@ describe('happypack', () => {
...
       ]),
     ])
     expect(config).toMatchSnapshot()
   })
 
   test('extractText', () => {
-    const config = createConfig.vanilla([
-      cssModules(),
+    const config = createConfig([
+      css(),
```
Comment (andywer) [inline test/index.test.js]: Should be `css.modules()`, right?

**Comment by diegohaz** [comment]
Thank you, @boxfoot and @andywer. Will have a look at this during the week.

**Comment by boxfoot** [comment]
OK!   Thanks for the feedback @andywer.   Was able to correct a bunch of subtle errors and fix all but one of the tests (although that test points to a significant remaining flaw I'm not sure how to tackle yet).

Many changes to properly handle passing options as an object instead of query string, and I made this the default format for output since that's the direction webpack is going.

Remaining test failure is from extractText().  The problem is that I'm not longer passing in `prevConfig` to the blocks within happypack, so extractText() can't recognize any rules to act on.

I'm not passing prevConfig because when I pass it in, happypack can no longer distinguish between rules created inside its inner blocks and rules created outside, and it ends up creating lots of duplicate plugins and incorrectly matching rules that it shouldn't.

So the question is:  is there a clever way to only correct the expected rules with happypack (for example, by executing `block(context, utils)(prevConfig)` and then diffing the result with `prevConfig` or is it OK to add a constraint that blocks within happypack should not depend on blocks outside of it....?

**Comment by boxfoot** [comment]
@diegohaz I think everything is now working, actually - would appreciate your eyes on that, and especially on the changes I made to the tests to make sure you agree with that.

Still needs an updated yarn.lock - I could install yarn and do that but if easy for someone else to add, that would be great.  Let me know.

Also, would you like me to squash this?  I can't tell your approach from the commit history.

**Comment by andywer** [comment]
@boxfoot Diego can just squash on merge if he wants to (GitHub introduced options to [create a merge commit, rebase or squash](https://github.com/blog/2243-rebase-and-merge-pull-requests) some time ago) :)

**Comment by andywer** [comment]
Sorry, I haven't had time to think about the [config patching issue](https://github.com/diegohaz/webpack-blocks-happypack/pull/18#issuecomment-342046363) in-depth yet. Might need to stick to diffing.

**Comment by boxfoot** [comment]
@andywer diffing worked, and wasn't actually that hard.

If this is a common use-case, it may be helpful for someone to create a library like `webpack-merge` called something like `webpack-diff` -- but it might be rare enough to just handle as needed.

**Comment by diegohaz** [comment]
Released as v0.2.0

Thank you, @boxfoot

**Comment by andywer** [comment]
Cool, guys! Ping me if you tweet about it and I will retweet :)

**Comment by boxfoot** [comment]
Awesome, thanks @andywer and @diegohaz !

**Pull request merged**